File: internal/registry/store.go
package registry

type RuntimeKnowledgeBase struct {
	ID                string         `json:"id"`
	Name              string         `json:"name"`
	StoreType         string         `json:"store_type"`
	StoreConfig       map[string]any `json:"store_config"`
	Enabled           bool           `json:"enabled"`
	DefaultSearchMode string         `json:"default_search_mode"`
	Indexing          map[string]any `json:"indexing"`
	Tags              []string       `json:"tags"`
}

type Store interface {
	List() ([]RuntimeKnowledgeBase, error)
	Save([]RuntimeKnowledgeBase) error
}

type MemoryStore struct {
	items []RuntimeKnowledgeBase
}

func NewMemoryStore(items []RuntimeKnowledgeBase) *MemoryStore {
	return &MemoryStore{items: items}
}

func (m *MemoryStore) List() ([]RuntimeKnowledgeBase, error) {
	out := make([]RuntimeKnowledgeBase, len(m.items))
	copy(out, m.items)
	return out, nil
}

func (m *MemoryStore) Save(items []RuntimeKnowledgeBase) error {
	m.items = make([]RuntimeKnowledgeBase, len(items))
	copy(m.items, items)
	return nil
}

File: knowledger/registry/store.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Protocol
import threading


@dataclass
class RuntimeKnowledgeBase:
    id: str
    name: str = ""
    store_type: str = ""
    store_config: dict[str, Any] = field(default_factory=dict)
    enabled: bool = True
    indexing: dict[str, Any] = field(default_factory=dict)
    tags: list[str] = field(default_factory=list)


class Store(Protocol):
    def list(self) -> list[RuntimeKnowledgeBase]: ...
    def save(self, items: list[RuntimeKnowledgeBase]) -> None: ...
    def version(self) -> str: ...


class MemoryStore:
    def __init__(self, items: list[RuntimeKnowledgeBase] | None = None) -> None:
        self._lock = threading.Lock()
        self._items: list[RuntimeKnowledgeBase] = list(items) if items else []
        self._version: int = 0

    def list(self) -> list[RuntimeKnowledgeBase]:
        with self._lock:
            return list(self._items)

    def save(self, items: list[RuntimeKnowledgeBase]) -> None:
        with self._lock:
            self._items = list(items)
            self._version += 1

    def version(self) -> str:
        with self._lock:
            return str(self._version)
